### PR TITLE
Change the implementation of most counting commands

### PR DESCRIPTION
**Pull request by Shredder121**
This uses the underlying map, very carefully, which in most cases are safe enough to call.

Note that the '`pom.xml`' that gradle produces has all its dependencies set to `runtime`.
So that means that trove isn't actually a transitive (compile time) dependency of JDA.
That sucks big time, but /shrug.

### CODE REVIEW

**Comment by Shredder121** [comment]
But the ShardsCommand has at least a few calls inlined, because it wants to do the counts incrementally.
What we could do is add them to a second list and then `JDAUtil.countAllGuilds` to only pay the hit once, and the same for `JDAUtil.countAllUniqueUsers`.

**Comment by Frederikam** [comment]
We could also choose to just have ;;mshards show non-unique users

**Comment by Shredder121** [comment]
Rebased, since 1) I forgot to add `JDAUtil` and 2) conflicts.

But do tell me, if the style is okay.
Or if you see more calls that could be changed.

**Comment by Shredder121** [comment]
Anyway, `;;shards` still has the parallelStream, but non-unique might also work.

**Comment by Shredder121** [comment]
Now only to have it in the feature configuration right?

**Comment by napstr** [comment]
Let's not user parallel Streams. I think those are always a bad choice.

**Comment by Frederikam** [comment]
>Let's not user parallel Streams. I think those are always a bad choice.

*Said the guy who used them to break the bot*

😛

**Comment by napstr** [comment]
I'm also the guy who tries to do every mistake exactly once.

**Comment by Shredder121** [comment]
inb4 JDA thinks it's fine:
```java
GuildImpl.java
@Override
public List<Emote> getEmotesByName(String name, boolean ignoreCase)
{
    Args.notNull(name, "name");
    return Collections.unmodifiableList(emotes.valueCollection().parallelStream()
            .filter(e ->
                    ignoreCase
                    ? StringUtils.equalsIgnoreCase(e.getName(), name)
                    : StringUtils.equals(e.getName(), name))
            .collect(Collectors.toList()));
}
```

File: FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java [hunk 1]
```diff
@@ -60,8 +61,9 @@ public void onInvoke(Guild guild, TextChannel channel, Member invoker, Message m
         LongOpenHashSet healthyUsers = new LongOpenHashSet(FredBoat.getExpectedUserCount());
         for (FredBoat fb : shards) {
             if (fb.getJda().getStatus() == JDA.Status.CONNECTED && !full) {
-                healthyGuilds += fb.getJda().getGuilds().size();
-                fb.getJda().getUsers().parallelStream().mapToLong(ISnowflake::getIdLong).forEach(healthyUsers::add);
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java]: This line is missing a replacement.
Need to add userIds to healthyUsers,

**Comment by napstr** [comment]
I wouldn't say everything JDA does works well at scale, so sadly it's not necessary a good place to get inspirations from.

File: FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java (hunk 1 above)
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java]: I thought we went the non-unique route.
So in this case we can update a regular long/int right?

If we are to dedupe them, let me know I have an idea.

Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java]: I thought we went the non-unique route.
So in this case we can update a regular long/int right?

If we are to dedupe them, let me know I have an idea.

File: FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java [hunk 2]
```diff
@@ -58,12 +57,12 @@ public void onInvoke(Guild guild, TextChannel channel, Member invoker, Message m
...
         int borkenShards = 0;
         int healthyGuilds = 0;
-        LongOpenHashSet healthyUsers = new LongOpenHashSet(FredBoat.getExpectedUserCount());
+        int healthyUsers = 0;
         for (FredBoat fb : shards) {
             if (fb.getJda().getStatus() == JDA.Status.CONNECTED && !full) {
                 healthyGuilds += fb.getGuildCount();
                 // casting to get the underlying map, this is safe because we only need the .size()
-                ((JDAImpl) fb.getJda()).getUserMap().size();
+                healthyUsers = ((JDAImpl) fb.getJda()).getUserMap().size();
```
Comment (napstr) [inline FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java]: This should be summing up.

File: FredBoat/feature_flags.properties [hunk 3]
```diff
@@ -31,4 +31,5 @@
 # example to turn a feature on (uncomment it):
 #RATE_LIMITER=false
 #CHATBOT=false
+#DATA_METHODS=false
```
Comment (napstr) [inline FredBoat/feature_flags.properties]: Fyi this file is almost purely decorative - it needs to be pushed and updated by hand to production.
It's good to keep it in sync with the existing flags, just keep in mind changing it in git changes nothing in production.

File: FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java (hunk 2 above)
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/command/maintenance/ShardsCommand.java]: `Actually store it correctly for real this time`
Very good catch >.<

File: FredBoat/feature_flags.properties (hunk 3 above)
Comment (Shredder121) [inline FredBoat/feature_flags.properties]: Yeah that would be annoying, I assumed you weren't working from inside the git directory.
Still, keeping it in sync is nice.

Comment (napstr) [inline FredBoat/feature_flags.properties]: Eventually it will move to the database and a GUI with buttons to click or something.

**Pull request merged**